File: 0042-trapping-rain-water/0042-trapping-rain-water.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        
        int res = 0, size = height.size();
        vector<int>leftMax(size), rightMax(size);
        
        leftMax[0] = height[0];
        rightMax[size - 1] = height[size - 1];
                           
        for(int i = 1; i < size; i++)
              leftMax[i] = max(leftMax[i - 1], height[i]);  
         for(int i = size - 2; i >= 0; i--)
              rightMax[i] = max(rightMax[i + 1], height[i]);    
                
        for(int i = 1; i < size - 1; i++)
            res+= min(leftMax[i], rightMax[i]) - height[i];
                           
        return res;
    }
};

File: 0042-trapping-rain-water/0042-trapping-rain-water.py
class Solution:
    def trap(self, height: List[int]) -> int:
        
        
        left, right = 0, len(height) - 1
        res = 0
        leftMax, rightMax = 0, 0
        
        while left < right:
            
            if height[left] < height[right]:            
                leftMax = max(leftMax, height[left])
                res+= leftMax - height[left]     
                left+=1
            else:              
                rightMax = max(rightMax, height[right])
                res+= rightMax - height[right] 
                right-=1
                
        return res
        
        